fl4health/strategies/feddg_ga_strategy.py

```diff
@@ -0,0 +1,382 @@
+
+            # initializing adjustment weights for this client if they don't exist yet
+            if cid not in self.adjustment_weights:
+                self.adjustment_weights[cid] = INITIAL_ADJUSTMENT_WEIGHT
+
+            # apply adjustment weights
+            weighted_client_parameters = parameters_to_ndarrays(result[1].parameters)
+            for i in range(len(weighted_client_parameters)):
+                weighted_client_parameters[i] = weighted_client_parameters[i] * self.adjustment_weights[cid]
+
+            # sum weighted parameters
+            if aggregated_results is None:
+                aggregated_results = weighted_client_parameters
+            else:
+                assert len(weighted_client_parameters) == len(aggregated_results)
+                for i in range(len(weighted_client_parameters)):
+                    aggregated_results[i] = aggregated_results[i] + weighted_client_parameters[i]
+
+        assert aggregated_results is not None
+        return aggregated_results
+
+    def update_weights_by_ga(self, server_round: int, cids: List[str]) -> None:
+        """
+        Update the self.adjustment_weights dictionary by calculating the new weights
+        based on the current server round, fit and evaluation metrics.
+
+        Args:
+            server_round: (int) the current server round.
+            cids: (List[str]) the list of client ids that participated in this round.
+        """
+        value_list = []
+        # calculating local vs global metric difference
+        for cid in cids:
+            assert (
+                cid in self.train_metrics and cid in self.evaluation_metrics
+            ), f"{cid} not in {self.train_metrics.keys()} or {self.evaluation_metrics.keys()}"
+
+            global_model_metric_value = self.evaluation_metrics[cid][self.fairness_metric.metric_name]
+            local_model_metric_value = self.train_metrics[cid][self.fairness_metric.metric_name]
+            assert isinstance(global_model_metric_value, float) and isinstance(local_model_metric_value, float)
+
+            value_list.append(global_model_metric_value - local_model_metric_value)
+
+        # calculating norm gap
+        value_list_ndarray = np.array(value_list)
+        max_value = np.max(np.abs(value_list_ndarray))
+
+        if max_value == 0:
+            log(
+                WARNING,
+                "Max value in metric diff list is 0. Adjustment weights will remain the same. "
+                + f"Value list: {value_list}",
+            )
+            norm_gap_list = np.zeros(len(value_list))
+        else:
+            norm_gap_list = value_list_ndarray / max_value
+
+        step_size = (1.0 / 3.0) * self.get_current_weight_step_size(server_round)
+
+        # updating weights
+        new_total_weight = 0.0
+        for i in range(len(cids)):
+            cid = cids[i]
+            self.adjustment_weights[cid] += self.fairness_metric.signal * norm_gap_list[i] * step_size
```

Comment:
Perhaps a note here on the signal value is worthwhile. Something like
```
For loss values, large and **positive** gaps imply worse generalization of global weights to local models. Therefore, we want to **increase** weight for these model parameters to improve generalization. So signal is positive. For accuracy, large **negative** gaps imply worse generalization. So the signal is -1.0, to increase weights for the associated model parameters.
```

Reply:
I'm adding this comment to `FairnessMetricType.signal_for_type` as well.